### PR TITLE
Use execution engine to instrument build scripts

### PR DESCRIPTION
**Pull request by asodja**
Fixes https://github.com/gradle/gradle/issues/27705

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

File: platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/BuildServices.kt [hunk 1]
```diff
@@ -123,6 +125,7 @@ object BuildServices {
             fileCollectionFactory,
             inputFingerprinter,
             gradlePropertiesController,
+            transformFactoryForLegacy
```
Comment (lptr) [inline platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/BuildServices.kt:128]: ❓ Why is this necessary?

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy [hunk 2]
```diff
@@ -409,10 +409,10 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jar(file)
         def classpath = DefaultClassPath.of(file)
-        def cachedFile = testDir.file("cached/8bccc8c51683456910b894e02dc7e915/thing.jar")
+        def cachedFile = testDir.file("cached/bc4f3e05266cfadaa3cd1f4484942739/thing.jar")
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy:412]: 🤔 Wonder how we could get rid of hashes like this in test code that need to be updated after some changes...

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java [hunk 3]
```diff
@@ -76,7 +89,10 @@ public Identity identify(Map<String, ValueSnapshot> identityInputs, Map<String,
     @Override
     public WorkOutput execute(ExecutionRequest executionRequest) {
         File workspace = executionRequest.getWorkspace();
-        compileTo(classesDir(workspace));
+        File output = compileTo(originalDir(workspace));
+        // We should instrument output of classes directory to the classes directory instead of a jar,
+        // but currently instrumentation classloader doesn't support that yet.
+        instrument(output, transformedJar(workspace));
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java:95]: :question: For me `transformed` was quite confusing in this context. Shall we call it `instrumented` instead?

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java [hunk 4]
```diff
@@ -104,12 +125,43 @@ public InputFingerprinter getInputFingerprinter() {
...
+            this.output = output;
+        }
+
+        public File getOutput() {
+            return output;
+        }
+
+        public File getOriginalDir() {
+            return originalDir;
+        }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java:165]: ❓ So `originalDir` is the compiled `.class` files _without_ instrumentation, right? And `output` is the same `.class` files, but _with_ instrumentation? If so, I think that's what we should call these methods.

I'm a bit confused why we need to carry both.

Could we make this clearer if compilation and instrumentation were separated? It feels like they shouldn't be handled in one place.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/BuildServices.kt (hunk 1 above)
Comment (asodja) [reply platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/BuildServices.kt:128]: We need it to create an object that does instrumentation.

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java (hunk 3 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java]: Changed and we now use `instrumented`

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java (hunk 4 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java]: I refactored code a bit and made it a bit more obvious. Now only Groovy's Unit of work  returns some additional data, that is more explicit: it returns metadata and instrumented jar, there is no more original dir.

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy (hunk 2 above)
Comment (asodja) [reply subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy:412]: Yes, this is annoying. I think when we remove all instrumentation with this custom cached classpath transformer we could also remove some if these test here and maybe get rid of that

Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy:412]: That's a good point. Let's make sure that we do remove them once we can, though my understanding is that we can't do it after merging this PR, right?

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClassCompiler.java [hunk 5]
```diff
@@ -110,16 +114,16 @@ public <T extends Script, M> CompiledScript<T, M> compile(
...
-        File outputDir = doCompile(target, templateId, sourceHashCode, remapped, classLoader, operation, verifier, scriptBaseClass);
+        GroovyScriptCompilationOutput output = doCompile(target, templateId, sourceHashCode, remapped, classLoader, operation, verifier, scriptBaseClass);
 
-        File genericClassesDir = classesDir(outputDir, operation);
-        File metadataDir = metadataDir(outputDir);
-        // TODO: Move instrumentation to the execution engine and remove the remapping or move remapping to the non-cacheable unit of work
-        ClassPath remappedClasses = remapClasses(genericClassesDir, remapped);
+        File instrumentedJar = output.getInstrumentedJar();
+        File metadataDir = output.getMetadataDir();
+        // TODO: Remove the remapping or move remapping to the non-cacheable unit of work?
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClassCompiler.java:121]: ❓ What's "the non-cacheable work"? Is this something we already have, or something to be potentially created later?

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java [hunk 6]
```diff
@@ -73,10 +81,20 @@ public Identity identify(Map<String, ValueSnapshot> identityInputs, Map<String,
         return () -> identity;
     }
 
+    @Override
+    public void visitOutputs(File workspace, OutputVisitor visitor) {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java:85]: :x: This should be annotated with `@OverridingMethodsMustInvokeSuper`.

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java [hunk 7]
```diff
@@ -34,36 +38,40 @@
 import static java.util.Objects.requireNonNull;
 
 /**
- * A base class that represents a work for compilation for Kotlin and Groovy scripts.
+ * A base class that represents a work for compilation for Kotlin and Groovy build scripts.
+ * This work unit first compiles the build script to a directory, and then instruments the directory for configuration cache and returns instrumented output.
  */
-public abstract class BuildScriptCompileUnitOfWork implements ImmutableUnitOfWork {
+public abstract class BuildScriptCompileAndInstrumentUnitOfWork implements ImmutableUnitOfWork {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java:44]: ❓ How about `BuildScriptCompilationAndInstrumentation`? We don't have the `UnitOfWork` suffix in `TaskExecution` nor in `TransformExecution`. And we use the noun form, not verbs, so "execution" instead of "execute".

Well, it's not entirely universal, but I think this is the pattern we want to follow, and change the rest of the names accordingly.

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy (hunk 2 above)
Comment (asodja) [reply subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy:412]: Yes, that is right. We need to clean up two more places to remove it

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java (hunk 6 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java:85]: Done

File: subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java (hunk 7 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileAndInstrumentUnitOfWork.java:44]: Sounds good, renamed

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClassCompiler.java (hunk 5 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClassCompiler.java:121]: Hehe, it's just means that we should not cache that unit of work. Changed `the non-cacheable unit of work` to `an uncached unit of work`. Does it make more sense now?

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy (hunk 2 above)
Comment (asodja) [reply subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy:412]: Actually looking at this test it looks like we will keep it even after we remove all custom classpath transformations, since we use that class also in other cases. 

Do you feel it's worth finding another solution now, or it can wait for another occasion?

Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy:412]: Nah, just annoyed that we have it. I think we done enough for now, let's think of a solution the next time we need to change something in this area.

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by asodja** [comment]
@bot-gradle test and merge
